File: backend/internal/core/model/question.go
package model

import (
	"github.com/google/uuid"
	"time"
)

type Question struct {
	ID   uuid.UUID     `json:"id"`
	Type *QuestionType `json:"type"`
	// Poll is only set if Type is QuestionTypePoll
	Poll      *Poll     `json:"poll"`
	Category  Category  `json:"category"`
	Author    User      `json:"author"`
	Title     string    `json:"title"`
	Body      *string   `json:"body"`
	Location  Location  `json:"location"`
	ImageURLs []string  `json:"image_urls"`
	Summary   *string   `json:"summary"`
	CreatedAt time.Time `json:"created_at"`
	EditedAt  time.Time `json:"edited_at"`
}

type Poll struct {
	Options       []PollOption `json:"options"`
	NumTotalVotes int          `json:"num_total_votes"`
	CreatedAt     time.Time    `json:"created_at"`
	ExpiresAt     time.Time    `json:"expires_at"`
}

type PollOption struct {
	Label    string `json:"label"`
	NumVotes int    `json:"num_votes"`
}

File: frontend/models/question.ts
import { User } from "@/models/user";
import {Location} from "@/models/location";
import {PageFilterType} from "@/services/axios-client";

export enum ContentType {
  POLL = "Poll",
  NONE = "None",
}

export enum Category {
  RESTAURANT = 'Restaurant',
  STORE = 'Store',
  TRANSPORTATION = 'Transportation',
  EVENT = 'Event',
  GENERAL = 'General',
}

export type QuestionContent = {
  type: ContentType,
  data: any,
}

export type PollOption = {
  id: string;
  is_selected: boolean;
  label: string;
  num_votes: number;
}

export type Poll = {
  id: string;
  question_id: string;
  options: PollOption[];
  num_total_votes: number;
  created_at: string;
  expired_at: string;
}

export type Question = {
  id: string;
  author: User;
  title: string;
  body?: string;
  category: Category;
  content: QuestionContent;
  location: Location;
  responses_amount: number;
  image_urls?: string[];
  is_owned: boolean;
  created_at: string;
  edited_at: string;
  expired_at: string;
};

// DTOs

// CREATE QUESTION
export type CreateQuestionReq = {
  title: string;
  body?: string;
  category: Category;
  location: Location;
  image_urls?: string[];
  duration: string;
  content_type: ContentType;
}

export type CreateQuestionRes = {
  question_id: string;
}

// GET QUESTION BY ID
export type GetQuestionByIdRes = {
  question: Question;
}

// CREATE POLL
export type CreatePollReq = {
  question_id: string;
  option_labels: string[];
}

export type CreatePollRes = {
  poll_id: string;
}

// VOTE POLL
export type VotePollReq = {
  question_id: string;
  poll_id: string;
  option_id?: string;
}

// GET QUESTIONS IN RADIUS FEED
export type GetQuestionsInRadiusFeedReq = {
  location: Location;
  radius_miles: number;
  limit: number;
  offset: number;
  page_filter_type: PageFilterType;
  page_filter_value: string;
}

export type GetQuestionsInRadiusFeedRes = {
  questions: Question[];
}
